### PR TITLE
[HttpFoundation] Validate/cast cookie expire time

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #... 
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 1]
```diff
@@ -63,15 +63,19 @@ public function __construct($name, $value = null, $expire = 0, $path = '/', $dom
         } elseif (!is_numeric($expire)) {
             $expire = strtotime($expire);
 
-            if (false === $expire || -1 === $expire) {
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: :+1: `-1` was returned by `strtotime` on PHP 5.1

Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: Prior to 5.1.0 actually ;)

**Comment by ro0NL** [comment]
Failing test unrelated.

File: src/Symfony/Component/HttpFoundation/Tests/CookieTest.php [hunk 2]
```diff
@@ -81,6 +89,13 @@ public function testGetExpiresTime()
         $this->assertEquals($expire, $cookie->getExpiresTime(), '->getExpiresTime() returns the expire date');
     }
 
+    public function testGetExpiresTimeIsCastedToInt()
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/CookieTest.php]: `testGetExpiresTimeIsCastToInt`

File: src/Symfony/Component/HttpFoundation/Tests/CookieTest.php [hunk 3]
```diff
@@ -81,6 +89,13 @@ public function testGetExpiresTime()
...
+    {
+        $cookie = new Cookie('foo', 'bar', 3600.9);
+
+        $this->assertEquals(3600, $cookie->getExpiresTime(), '->getExpiresTime() returns the expire date as an integer');
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/CookieTest.php]: `assertSame`

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/Tests/CookieTest.php (hunk 3 above)
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Tests/CookieTest.php]: Nice catch!

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 4]
```diff
@@ -112,15 +112,19 @@ public function __construct($name, $value = null, $expire = 0, $path = '/', $dom
...
+            if (false === $expire) {
                 throw new \InvalidArgumentException('The cookie expiration time is not valid.');
             }
         }
 
+        if (0 > $expire) {
+            throw new \InvalidArgumentException('The cookie expiration time must be a positive number value.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: This is a bc break if anyone is doing this for some reason.
We could deprecate, but the behavior is just fine when negative numbers are used here.
Instead, I'd rather fix getMaxAge and turn `0 !== $this->expire` into `0 >= $this->expire`.
and merge this in a lower branch

Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: I think we should indeed settle with `>=`. Nice one.. didnt thought of it :+1:

**Comment by nicolas-grekas** [comment]
:+1:

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 5]
```diff
@@ -56,15 +56,15 @@ public function __construct($name, $value = null, $expire = 0, $path = '/', $dom
...
+            if (false === $expire) {
                 throw new \InvalidArgumentException('The cookie expiration time is not valid.');
             }
         }
 
         $this->name = $name;
         $this->value = $value;
         $this->domain = $domain;
-        $this->expire = $expire;
+        $this->expire = 0 < $expire ? (int) $expire : 0;
```
Comment (javiereguiluz) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: `$this->expire = max(0, (int) $expire)`  ?

Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: Not sure the extra function call is really worth it. Tend to keep it as is.

**Comment by fabpot** [comment]
Thank you @ro0NL.

**Pull request merged**